File: ios-app/CalendarioCiclismo/Views/Components/TVBadge.swift
import SwiftUI

/// Badge de estado de TV con hora de emisión si está disponible.
/// Clicable cuando hay una URL de emisión o de live texto disponible
/// (misma prioridad que la web: YouTube > otras redes sociales > RTVE.es > resto).
struct TVBadge: View {
    let tvStatus: String?
    let broadcasts: [Broadcast]
    var neutralStartTimeUtc: String? = nil
    var liveTextUrl: String? = nil
    @Environment(\.accessibilityShowButtonShapes) private var showButtonShapes
    @Environment(\.openURL) private var openURL
    @State private var regionService = RegionService.shared

    private var isHighContrast: Bool { showButtonShapes }

    /// Broadcasts visibles para la región del usuario. TODO el badge (label, icono,
    /// "hay TV", hora de referencia y enlace) se calcula sobre esta lista, no sobre
    /// la cruda: si no, un usuario de España vería "TV 14:00" de una emisión solo
    /// de Bélgica aunque el enlace estuviera (correctamente) suprimido. La web ya
    /// pre-filtra los broadcasts antes de pintar el badge; aquí filtramos en sitio.
    private var regionBroadcasts: [Broadcast] {
        RaceLogic.filterBroadcastsByRegion(
            broadcasts,
            allowedGroups: regionService.current.allowedBroadcastGroups
        )
    }

    /// True si HABÍA emisiones pero NINGUNA es accesible en la región del usuario
    /// (todas filtradas). En ese caso el usuario no puede ver nada → no mostramos el
    /// badge "TV" genérico que vendría de `tvStatus='confirmed'`. (Sin emisiones en
    /// absoluto, el badge sigue saliendo de `tvStatus`: cobertura sin canal aún.)
    private var regionBlocked: Bool { !broadcasts.isEmpty && regionBroadcasts.isEmpty }

    /// True si se debe mostrar "Live texto" en vez del estado TV.
    private var showLiveText: Bool {
        guard let url = liveTextUrl, !url.isEmpty else { return false }
        let status = tvStatus ?? ""
        return status == "none" || status == "unavailable_es" || status == "pending"
            || (status.isEmpty && regionBroadcasts.isEmpty)
    }

    /// True si la emisión ya está EN DIRECTO (su hora de inicio ya pasó).
    private func isBroadcastLive(_ b: Broadcast) -> Bool {
        guard let ts = b.startTimeUtc, let date = DateFormatting.parseISO(ts) else { return false }
        return Date() >= date
    }

    /// Broadcast seleccionado para el enlace: filtrado por región. Una emisión YA EN
    /// DIRECTO gana SIEMPRE a una que aún no ha empezado, aunque esta última sea de
    /// mayor tier (p. ej. Eurosport ya emitiendo vs RTVE por empezar → enlaza a
    /// Eurosport, lo accesible AHORA). A igualdad de estado manda el tier (YouTube >
    /// otras redes sociales > RTVE.es > resto; Eurosport / HBO Max son "una cadena
    /// más") y luego el sortOrder. Sin ninguna en directo se conserva el tier puro.
    private var selectedBroadcast: Broadcast? {
        if showLiveText { return nil }
        return regionBroadcasts
            .filter { b in b.url.map { !$0.isEmpty } == true }
            .sorted {
                let l0 = isBroadcastLive($0) ? 0 : 1, l1 = isBroadcastLive($1) ? 0 : 1
                if l0 != l1 { return l0 < l1 }
                let t0 = RaceLogic.broadcastLinkPriority($0.url), t1 = RaceLogic.broadcastLinkPriority($1.url)
                if t0 != t1 { return t0 < t1 }
                return ($0.sortOrder ?? 0) < ($1.sortOrder ?? 0)
            }
            .first
    }

    /// Broadcast con hora más temprana (fallback cuando no hay broadcast con URL).
    private var earliestTimedBroadcast: Broadcast? {
        regionBroadcasts
            .compactMap { b -> (Broadcast, Double)? in
                guard let ts = b.startTimeUtc else { return nil }
                return (b, DateFormatting.timestampToSeconds(ts) ?? 0)
            }
            .sorted { $0.1 < $1.1 }
            .first
            .map(\.0)
    }

    /// True si la carrera ya ha comenzado (neutralStart alcanzado).
    private var raceStarted: Bool {
        guard let ts = neutralStartTimeUtc,
              let date = DateFormatting.parseISO(ts) else { return false }
        return Date() >= date
    }

    /// True si la emisión de referencia (la más temprana accesible) ya ha comenzado.
    private var broadcastStarted: Bool {
        guard let ts = earliestTimedBroadcast?.startTimeUtc, let date = DateFormatting.parseISO(ts) else { return false }
        return Date() >= date
    }

    /// True si el badge debe mostrarse como "Live" (verde, retransmisión en curso).
    private var showLive: Bool {
        tvStatus == "confirmed_time" && broadcastStarted
    }

    /// True si hay que mostrar el chip "Live texto" JUNTO al badge de TV: la carrera ya empezó
    /// pero la emisión de referencia aún no ha comenzado (hora futura). Paridad con la web
    /// (`liveTextAlongside` en `tvBadgeCard`, js/app.js). No aplica cuando `showLiveText` ya
    /// sustituye al badge de TV (casos sin TV: none / pending / sin broadcasts).
    private var showLiveTextAlongside: Bool {
        guard !showLiveText,
              let url = liveTextUrl, !url.isEmpty,
              raceStarted else { return false }
        guard let ts = earliestTimedBroadcast?.startTimeUtc, let date = DateFormatting.parseISO(ts) else { return false }
        return Date() < date
    }

    private var label: String {
        // Prioridad: live texto cuando no hay TV
        if showLiveText { return LocaleService.t("Live texto", "Live text") }

        // Retransmisión en vivo: la hora de inicio ya ha pasado
        if showLive { return "Live" }

        // Comprobar estados "sin TV" ANTES de broadcasts (misma lógica que web)
        switch tvStatus {
        case "none": return LocaleService.t("Sin TV", "No TV")
        case "unavailable_es":
            // Ese estado solo importa al usuario en España (en EN la app sirve
            // a europeos sin restricción geográfica). Ocultar devolviendo "".
            return LocaleService.isEnglish ? "" : "No TV España"
        case "pending": return LocaleService.t("Sin confirmar", "Unconfirmed")
        default: break
        }

        // Hora del badge = la emisión accesible que ANTES empieza (independiente del
        // enlace): si una emisión global (ALL) empieza antes que las de tu grupo, su
        // hora manda. El enlace (tap) sigue por prioridad de tier vía `tappableUrl`.
        let refBroadcast = earliestTimedBroadcast
        if let ref = refBroadcast, let broadcastTs = ref.startTimeUtc {
            if let neutralTs = neutralStartTimeUtc,
               let neutralDate = DateFormatting.parseISO(neutralTs),
               let broadcastDate = DateFormatting.parseISO(broadcastTs),
               broadcastDate <= neutralDate {
                return LocaleService.t("Íntegra", "Full Race")
            }
            // La web muestra solo la hora: mantener la misma etiqueta en las apps.
            if let time = ref.startTimeLocal { return time }
        }
        if !regionBroadcasts.isEmpty { return "TV" }

        // TV confirmada pero toda fuera de la región del usuario → sin badge.
        if regionBlocked { return "" }

        if tvStatus == "confirmed" { return "TV" }

        // Sin tvStatus y sin broadcasts
        return ""
    }

    private var iconName: String {
        if showLiveText { return "text.bubble" }
        if !regionBroadcasts.isEmpty || tvStatus == "confirmed" { return "tv" }
        if tvStatus == "pending" { return "tv" }
        if tvStatus == "unavailable_es" { return "tv.slash" }
        if tvStatus == "none" { return "tv.slash" }
        return "tv.slash"
    }

    /// URL a abrir al pulsar el badge (derivada de `selectedBroadcast`).
    private var tappableUrl: URL? {
        if showLiveText, let urlStr = liveTextUrl, !urlStr.isEmpty {
            return URL(string: urlStr)
        }
        guard let urlStr = selectedBroadcast?.url else { return nil }
        return URL(string: urlStr)
    }

    @ViewBuilder
    private func badgeShape(text: String, iconName: String, colors: AppTheme.BadgeColor, accessibilityLabel: String) -> some View {
        HStack(spacing: 3) {
            Image(systemName: iconName)
                .font(.system(size: 9))
            Text(text)
                .font(.caption2)
                .fontWeight(.medium)
                .textCase(.uppercase)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 3)
        .background(colors.background)
        .foregroundStyle(colors.foreground)
        .clipShape(RoundedRectangle(cornerRadius: 3))
        .overlay(
            isHighContrast
                ? RoundedRectangle(cornerRadius: 3).strokeBorder(colors.foreground, lineWidth: 1)
                : nil
        )
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityLabel)
    }

    /// Chip independiente de "Live texto" (icono burbuja + enlace). Verde si la carrera ya empezó,
    /// azul ("pre") si aún no. Se usa tanto cuando sustituye al badge de TV (`showLiveText`) como
    /// cuando acompaña al badge de TV (`showLiveTextAlongside`).
    @ViewBuilder
    private var liveTextChip: some View {
        let key = raceStarted ? "livetext" : "livetext_pre"
        let colors = AppTheme.tvStatusColor(for: key, hasBroadcasts: false, highContrast: isHighContrast)
        let shape = badgeShape(
            text: LocaleService.t("Live texto", "Live text"),
            iconName: "text.bubble",
            colors: colors,
            accessibilityLabel: LocaleService.t("Live texto disponible", "Live text available")
        )
        if let urlStr = liveTextUrl, let url = URL(string: urlStr) {
            Button { openURL(url) } label: { shape }
                .buttonStyle(.plain)
        } else {
            shape
        }
    }

    var body: some View {
        if showLiveText {
            // Caso sin TV con live texto: el chip de live texto SUSTITUYE al badge de TV.
            liveTextChip
        } else {
            let text = label
            if !text.isEmpty {
                let hasBroadcasts = !regionBroadcasts.isEmpty
                let colors: AppTheme.BadgeColor = showLive
                    ? AppTheme.tvStatusColor(for: "tv_live", hasBroadcasts: hasBroadcasts, highContrast: isHighContrast)
                    : AppTheme.tvStatusColor(for: tvStatus, hasBroadcasts: hasBroadcasts, highContrast: isHighContrast)
                let accessibilityLabel = showLive
                    ? "Live"
                    : (AccessibilityTVStatus.description(tvStatus: tvStatus, broadcasts: regionBroadcasts) ?? text)
                let tvBadge = badgeShape(text: text, iconName: iconName, colors: colors, accessibilityLabel: accessibilityLabel)

                // Carrera empezada y TV aún en reposo → "Live texto" junto al badge de TV (paridad web).
                HStack(spacing: 4) {
                    if let url = tappableUrl {
                        Button { openURL(url) } label: { tvBadge }
                            .buttonStyle(.plain)
                    } else {
                        tvBadge
                    }
                    if showLiveTextAlongside {
                        liveTextChip
                    }
                }
            }
        }
    }
}

File: android-app/app/src/main/java/app/calendariociclismo/android/ui/components/TVBadge.kt
package app.calendariociclismo.android.ui.components

import android.content.Context
import android.content.Intent
import androidx.browser.customtabs.CustomTabsIntent
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChatBubbleOutline
import androidx.compose.material.icons.filled.Tv
import androidx.compose.material.icons.filled.TvOff
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.runtime.collectAsState
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.core.net.toUri
import app.calendariociclismo.android.R
import app.calendariociclismo.android.data.model.Broadcast
import app.calendariociclismo.android.data.prefs.RegionPreference
import app.calendariociclismo.android.ui.rememberApp
import app.calendariociclismo.android.ui.theme.tvStatusBadgeColor
import app.calendariociclismo.android.util.DateFormatting
import app.calendariociclismo.android.util.LocaleHolder
import app.calendariociclismo.android.util.RaceLogic

/**
 * Badge de estado de TV con hora de emisión si está disponible.
 * Clicable cuando hay URL de emisión o de live texto. La URL se elige filtrando
 * primero por la región del usuario y respetando estrictamente el sortOrder del admin.
 *
 * Port de `ios-app/.../Views/Components/TVBadge.swift`, con la misma lógica
 * de prioridades: live_text → none → unavailable_es → pending → broadcasts
 * con hora → confirmed.
 */
@Composable
fun TVBadge(
    tvStatus: String?,
    broadcasts: List<Broadcast>,
    neutralStartTimeUtc: String? = null,
    liveTextUrl: String? = null,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val app = rememberApp()
    val regionPref by app.preferences.regionPreference.collectAsState(initial = RegionPreference.SPAIN)

    // Todo el badge (label, icono, "hay TV", hora de referencia y enlace) se calcula
    // sobre los broadcasts visibles para la región del usuario, NO sobre la lista
    // cruda: si no, un usuario de España vería "TV 14:00" de una emisión solo de
    // Bélgica aunque el enlace estuviera (correctamente) suprimido. La web ya
    // pre-filtra los broadcasts antes de pintar el badge.
    val regionBroadcasts = RaceLogic.filterBroadcastsByRegion(broadcasts, regionPref.allowedBroadcastGroups)
    val hasBroadcasts = regionBroadcasts.isNotEmpty()
    // HABÍA emisiones pero NINGUNA accesible en la región del usuario (todas filtradas)
    // → no mostramos el badge "TV" genérico que vendría de tvStatus='confirmed'. (Sin
    // emisiones en absoluto, el badge sigue saliendo de tvStatus: cobertura sin canal aún.)
    val regionBlocked = broadcasts.isNotEmpty() && regionBroadcasts.isEmpty()

    val isLiveText = !liveTextUrl.isNullOrEmpty() && run {
        val s = tvStatus.orEmpty()
        s == "none" || s == "unavailable_es" || s == "pending" || (s.isEmpty() && !hasBroadcasts)
    }

    val nowSec = System.currentTimeMillis() / 1000.0
    val raceStarted = neutralStartTimeUtc
        ?.let { DateFormatting.timestampToSeconds(it) }
        ?.let { it <= nowSec } ?: false

    // Broadcast seleccionado para el enlace: región del usuario. Una emisión YA EN
    // DIRECTO (su hora de inicio ya pasó) gana SIEMPRE a una que aún no ha empezado,
    // aunque esta última sea de mayor tier (p. ej. Eurosport ya emitiendo vs RTVE por
    // empezar → enlaza a Eurosport, lo accesible AHORA). A igualdad de estado manda el
    // tier (YouTube > otras redes sociales > RTVE.es > resto; Eurosport / HBO Max son
    // "una cadena más") y luego el sortOrder. Sin ninguna en directo se conserva el
    // tier puro. Espejo iOS/web.
    val isBroadcastLive: (Broadcast) -> Boolean = { b ->
        b.startTimeUtc?.let { DateFormatting.timestampToSeconds(it) }?.let { it <= nowSec } ?: false
    }
    val selectedBroadcast: Broadcast? = if (isLiveText) null else {
        regionBroadcasts
            .filter { !it.url.isNullOrEmpty() }
            .sortedWith(
                compareBy(
                    { if (isBroadcastLive(it)) 0 else 1 },
                    { RaceLogic.broadcastLinkPriority(it.url) },
                    { it.sortOrder },
                )
            )
            .firstOrNull()
    }

    // Hora de referencia del badge = la emisión accesible que ANTES empieza (no la del
    // enlace): si una emisión global (ALL) empieza antes que las de tu grupo, su hora
    // manda. El enlace (tap) sigue por prioridad de tier vía selectedBroadcast.
    val refBroadcastSec = regionBroadcasts
        .mapNotNull { it.startTimeUtc?.let { ts -> DateFormatting.timestampToSeconds(ts) } }
        .minOrNull()
    val broadcastStarted = refBroadcastSec != null && refBroadcastSec <= nowSec
    val showLive = tvStatus == "confirmed_time" && broadcastStarted

    // Mostrar "Live texto" JUNTO al badge de TV: la carrera ya empezó pero la emisión de
    // referencia aún no ha comenzado (hora futura). Paridad con la web (`liveTextAlongside`).
    val showLiveTextAlongside = !isLiveText &&
        !liveTextUrl.isNullOrEmpty() &&
        raceStarted &&
        refBroadcastSec != null &&
        refBroadcastSec > nowSec

    // Caso sin TV con live texto: el chip de live texto SUSTITUYE al badge de TV.
    if (isLiveText) {
        LiveTextChip(
            liveTextUrl = liveTextUrl,
            raceStarted = raceStarted,
            context = context,
            modifier = modifier,
        )
        return
    }

    val label = buildLabel(
        tvStatus = tvStatus,
        broadcasts = regionBroadcasts,
        neutralStartTimeUtc = neutralStartTimeUtc,
        showLive = showLive,
        selectedBroadcast = selectedBroadcast,
        regionBlocked = regionBlocked,
    ) ?: return

    val colors = tvStatusBadgeColor(
        status = tvStatus,
        hasBroadcasts = hasBroadcasts,
        isTvLive = showLive,
    )
    val icon = when {
        hasBroadcasts || tvStatus == "confirmed" || tvStatus == "pending" -> Icons.Filled.Tv
        else -> Icons.Filled.TvOff
    }

    // URL a abrir al pulsar: la del broadcast seleccionado.
    val tappableUrl: String? = selectedBroadcast?.url

    val clickModifier = tappableUrl?.let { url ->
        Modifier.clickable(role = Role.Button) { openTvUrl(context, url) }
    } ?: Modifier

    Row(
        modifier = modifier
            .then(clickModifier)
            .background(colors.background, RoundedCornerShape(3))
            .padding(horizontal = 8.dp, vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(3.dp),
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = colors.foreground,
            modifier = Modifier.size(10.dp),
        )
        Text(
            text = label.uppercase(LocaleHolder.currentState),
            style = MaterialTheme.typography.labelSmall,
            fontWeight = FontWeight.Medium,
            color = colors.foreground,
        )
    }

    // Chip de "Live texto" como hermano del badge de TV. El FlowRow padre lo separa con el
    // mismo spacing (4.dp) que el resto de badges.
    if (showLiveTextAlongside) {
        LiveTextChip(
            liveTextUrl = liveTextUrl,
            raceStarted = raceStarted,
            context = context,
        )
    }
}

/**
 * Chip independiente de "Live texto" (icono burbuja + enlace). Verde si la carrera ya empezó,
 * azul ("pre") si aún no. Se usa tanto cuando sustituye al badge de TV (sin TV) como cuando
 * acompaña al badge de TV (carrera empezada y TV aún en reposo).
 */
@Composable
private fun LiveTextChip(
    liveTextUrl: String?,
    raceStarted: Boolean,
    context: Context,
    modifier: Modifier = Modifier,
) {
    val colors = tvStatusBadgeColor(
        status = null,
        hasBroadcasts = false,
        isLiveText = raceStarted,
        isLiveTextPre = !raceStarted,
    )
    val clickModifier = liveTextUrl?.takeUnless { it.isEmpty() }?.let { url ->
        Modifier.clickable(role = Role.Button) { openTvUrl(context, url) }
    } ?: Modifier
    Row(
        modifier = modifier
            .then(clickModifier)
            .background(colors.background, RoundedCornerShape(3))
            .padding(horizontal = 8.dp, vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(3.dp),
    ) {
        Icon(
            imageVector = Icons.Filled.ChatBubbleOutline,
            contentDescription = stringResource(R.string.tv_badge_live_text),
            tint = colors.foreground,
            modifier = Modifier.size(10.dp),
        )
        Text(
            text = stringResource(R.string.tv_badge_live_text).uppercase(LocaleHolder.currentState),
            style = MaterialTheme.typography.labelSmall,
            fontWeight = FontWeight.Medium,
            color = colors.foreground,
        )
    }
}

@Composable
private fun buildLabel(
    tvStatus: String?,
    broadcasts: List<Broadcast>,
    neutralStartTimeUtc: String?,
    showLive: Boolean,
    selectedBroadcast: Broadcast?,
    regionBlocked: Boolean = false,
): String? {
    // Retransmisión en vivo: la hora de inicio ya ha pasado.
    if (showLive) return stringResource(R.string.tv_badge_live)

    // Estados "sin TV" se priorizan sobre los broadcasts (misma lógica que web/iOS).
    when (tvStatus) {
        "none" -> return stringResource(R.string.tv_badge_no_tv)
        "unavailable_es" -> {
            // Solo aplica a usuarios en España. En EN (Premium) la app sirve
            // a europeos sin restricción geográfica → ocultar el badge.
            if (LocaleHolder.current.language == "en") return null
            return stringResource(R.string.tv_badge_no_tv_spain)
        }
        "pending" -> return stringResource(R.string.tv_badge_pending)
    }

    // Hora del badge = la emisión accesible que ANTES empieza (independiente del
    // enlace): si una emisión global (ALL) empieza antes que las de tu grupo, su hora
    // manda. El enlace (tap) sigue por prioridad de tier vía selectedBroadcast.
    val first = broadcasts
        .mapNotNull { b ->
            val ts = b.startTimeUtc ?: return@mapNotNull null
            val seconds = DateFormatting.timestampToSeconds(ts) ?: return@mapNotNull null
            b to seconds
        }
        .minByOrNull { it.second }
        ?.first

    if (first != null) {
        val broadcastTs = first.startTimeUtc
        if (broadcastTs != null && neutralStartTimeUtc != null) {
            val neutralSec = DateFormatting.timestampToSeconds(neutralStartTimeUtc)
            val broadcastSec = DateFormatting.timestampToSeconds(broadcastTs)
            if (neutralSec != null && broadcastSec != null && broadcastSec <= neutralSec) {
                return stringResource(R.string.tv_badge_full)
            }
        }
        // La web muestra solo la hora: mantener la misma etiqueta en las apps.
        first.startTimeLocal?.let { return it }
    }
    if (broadcasts.isNotEmpty()) return stringResource(R.string.tv_badge_tv)
    // TV confirmada pero toda fuera de la región del usuario → sin badge.
    if (regionBlocked) return null
    if (tvStatus == "confirmed") return stringResource(R.string.tv_badge_tv)
    return null
}

/** YouTube, HBO Max y X abren en su app nativa; el resto con Custom Tabs. */
private fun openTvUrl(context: Context, url: String) {
    val lower = url.lowercase()
    val prefersNative = lower.contains("youtube.com") || lower.contains("youtu.be") ||
        lower.contains("hbomax.com") || lower.contains("play.max.com") ||
        lower.contains("x.com") || lower.contains("twitter.com")
    runCatching {
        val uri = url.toUri()
        if (prefersNative) {
            context.startActivity(
                Intent(Intent.ACTION_VIEW, uri).apply {
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                }
            )
        } else {
            CustomTabsIntent.Builder().setShowTitle(true).build().launchUrl(context, uri)
        }
    }
}
